### ISSUE TITLE
Device nodes could get wrong permissions

### ISSUE THREAD

**Issue by ymgve** [qbarnes/catweasel-cw#1]
There is no explicit file permissions set when the install script runs `mknod`, which means some operating systens might not set the correct write permissions. If someone only uses the tool for simple reading this isn't an issue, but if someone wants to write to disks or set custom disk options, this will fail.

**Comment by qbarnes** [comment]
Good to see someone actually _has_ a Catweasel out there and is using it!

What permissions did you end up with on `/dev/cw?raw[01]` and what would you want?

As a workaround, the person running the script could set `umask` before running to get the permissions they wanted.

It could be also argued (and quite successfully!) that creating device nodes with `mknod` in a script is deprecated and should be done by using udev .rules file.

**Comment by ymgve** [comment]
Thanks for your patch to make it run on modern Linux - I was soo deep in reading about kernel timers and was like 60% done when I noticed that your version already had fixed this.

This is on Debian 11.6.0, and the permissions ended up with `rw-r--r--` owned by root, so no user could perform write operations on the device. I'm not really that familiar with Linux systems, so not sure how it would be implemented with udev, so for now I just do a mknod call which sets the correct permissions (also have to do this anyway after boot because I don't know how to make udev keep the devices permanent)

**Comment by qbarnes** [comment]
> I was soo deep in reading about kernel timers and was like 60% done when I noticed that your version already had fixed this.

Yep.  I'm an old device driver writer with a Catweasel MK4.  I did core kernel work on Linux, but most of my h/w device driver work has been on Solaris and BSD though.

As far as I can tell, the CW code's been abandoned, and I wanted it fixed.  Hopefully, I did it right.  I also wanted to create an git archive of CW for others looking for the code as well and wanting to do something with it.

If you ever want to get rid of your Catweasel someday, please let me make an offer on it.  My email is in my GitHub profile.

> ... permissions ended up with rw-r--r-- owned by root, so no user could perform write operations on the device ...

You should just be able to run `(umask 0; bash tools/install.bash)` as a workaround (but I didn't try it).  If you do that, make sure to leave the parentheses around the command when you do, so it doesn't change your current shell's `umask`.

> ... so not sure how it would be implemented with udev ...

I've been trying to create a udev rule for CW for a few hours, now but not having any luck and I don't know why yet.  When I do, I'll let you know.

Yes, a udev rule would have the effect of making the `/dev` entries permanent surviving reboots.  (Technically, they'd be recreated for you at reboot time.)

Kind of curious though why you're tinkering with CW.   The tool `cw2dmk` is much more robust and does a lot of the work of reading and recovering floppies in user space.  However, it's generally limited to certain floppy formats.

There's the more current approach of reading old floppies these days is with using a Greaseweazle rather than a Catweasel unless the reader was in need of the features of `cw2dmk`.  At present, the Greaseweazle software has a lot of breadth, but not a lot of depth yet, which may be good or bad for what you're trying to do.  Its software is written in Python, but its developer found that Python isn't real-time enough for doing floppy reads and writes and is thinking of rewriting it in a compilable language.

**Comment by qbarnes** [comment]
> Yes, a udev rule would have the effect of making the /dev entries permanent surviving reboots. (Technically, they'd be recreated for you at reboot time.)

Oops.  This was misleading.  I should have said that the `/dev` entries would be created when the module was loaded and removed when the module was unloaded.

**Comment by ymgve** [comment]
> Kind of curious though why you're tinkering with CW. The tool cw2dmk is much more robust and does a lot of the work of reading and recovering floppies in user space. However, it's generally limited to certain floppy formats.

That "limited to certain formats" is the reason - I use it on a mix of Amiga and Atari disks, and as far as I can see cw2dmk can't do the Amiga format. Though I've mostly switched over to just interfacing with the device node directly with my own python scripts for more flexibility.

**Comment by qbarnes** [comment]
Yes, `cw2dmk` mainly supports floppy formats that can have their contents represented as [DMK](http://cpmarchives.classiccmp.org/trs80/mirrors/www.discover-net.net/~dmkeil/coco/cocotech.htm) files.  I know that format definitely doesn't work for Amiga with its streaming track format.  I suspect the same for Atari, but not sure.

I just pushed a branch (`dev_create`) that has code that when the driver is loaded, it'll automatically create the device nodes for controllers found and with 0666 permissions.  (And get rid of them on `rmmod cw`.)  For example, these show up on my system when loaded:
```
$ sudo insmod cw.ko
$ ls -l /dev/cw*
crw-rw-rw-. 1 root root 120, 31 Mar 19 18:22 /dev/cw0raw0
crw-rw-rw-. 1 root root 120, 63 Mar 19 18:22 /dev/cw0raw1
```

This is a quick and dirty port to do the job.  It needs more work to make it friendly to non-Linux systems, but in the mean time, I hope it'll do the job for you.  Let me know if it works for you.

**Comment by qbarnes** [comment]
@ymgve, did the `dev_create` branch address your issue?

**Comment by ymgve** [comment]
I got this compile error with the new code:

```
/home/user/cw2/catweasel-cw/src/driver/driver.c: In function ‘cw_driver_module_exit’:
/home/user/cw2/catweasel-cw/src/driver/driver.c:182:4: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
  182 |    for (int f = 0; f < CW_NR_FLOPPIES_PER_CONTROLLER; ++f)
      |    ^~~
/home/user/cw2/catweasel-cw/src/driver/driver.c:182:4: note: use option ‘-std=c99’, ‘-std=gnu99’, ‘-std=c11’ or ‘-std=gnu11’ to compile your code
/home/user/cw2/catweasel-cw/src/driver/driver.c: In function ‘cw_driver_module_init’:
/home/user/cw2/catweasel-cw/src/driver/driver.c:270:3: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
  270 |   for (int f = 0; f < CW_NR_FLOPPIES_PER_CONTROLLER; ++f)
      |   ^~~
```

**Comment by ymgve** [comment]
I did a quick fix with variable declarations outside the loops and the code seems to work, devices are now properly created at startup.

**Comment by qbarnes** [comment]
You're right!  The code I wrote and tested was using a newer Linux kernel.  I didn't spin up an old VM to verify it would compile for you on Debian 11.6, which I think is running a 5.10 kernel.

Linus bumped the C standard for the kernel from C89 to C11 starting with the [5.18 kernel](https://kernelnewbies.org/Linux_5.18#Switch_to_C11).  C99 and later allows the use of variable definitions with the for statement.

I pushed a fix on the `dev_create` branch to support the older kernels.  Let me know if it works for you.
